### ISSUE TITLE
Provide a mechanism to suppress certain Terraform Validate failures

### ISSUE THREAD

**Issue by kaizoku0506** [hashicorp/terraform#21092]
### Current Terraform Version
```
0.11.8
```

### Use-cases

We are adding `terraform validate` and `terraform fmt` to all commits using a Git pre-commit hook.  However, there are some errors in TF that we expect and wish to set to "warn" or simply to suppress (provider versions, for instance).

In our case, all users develop on local workstations and push to Git.  Upon opening an MR, Jenkins pulls the correct provider versions and module versions from artifact storage before executing `terraform plan`.  We want to be able to use `terraform validate` to ensure they are providing all required inputs to their specified modules, but don't necessarily want to check for updated provider versions (given we often version lock) nor do we want to throw errors for providers which are not publicly available or installed on the workstation (given Jenkins has access to these, but many users won't have them installed locally).

### Attempted Solutions
At the moment, we are ignoring failures forcibly by passing this:

```
terraform validate /$(git rev-parse --show-toplevel)/${i} || true;
```

We tried intercepting errors and outputting our own warning messages depending upon the error, but found that the output of `terraform validate` couldn't be redirected in Bash and so output was not being evaluated as expected.

```
project_validate() {
  echo "${BOLD}Beginning linting with Terraform Validate...${NOFORMAT}";

  for i in ${projectlist_sorted[@]}; do
    echo "Running linting for project ${i}...";
    echo "";
    fail=0;

    while read LINE; do
      if [[ ${LINE} == *"Error:"* ]]; then
        echo "${BOLD}Linting uncovered the following issues for Project ${i}:${NOFORMAT}";
        echo "";
      elif [[ ${LINE} == *"provider."* ]]; then
        echo "${YELLOW}${LINE} - This may be a false positive.${NOFORMAT}";
      elif [[ ${LINE} == *"missing required argument"* ]]; then
        echo "${RED}${LINE}${NOFORMAT}";
        fail=1;
      else
        echo "${LINE}";
      fi
    done < <(terraform validate /$(git rev-parse --show-toplevel)/${i})
  done

  echo "";

  if [[ ${fail} == 0 ]]; then
    echo "${GREEN}Finished linting with Terraform Validate.${NOFORMAT}";
  else
    echo "${RED}Linting uncovered issues which must be corrected before commit.${NOFORMAT}";
    finish;
}
```

### Proposal
Provide multiple tiers / severities of errors (warn, err, fatal, etc.).  Provide a mechanism by which users can add a specific error / class of error to a given tier / severity (for instance, "missing required argument" == FATAL, but "provider.* new or changed plugin executable" == WARN, and "provider.* no suitable version installed" == IGNORE).  This would provide flexibility for users to define test criteria on local developer systems while avoiding unnecessary white noise.

**Comment by apparentlymart** [comment]
Hi @kaizoku0506! Thanks for sharing this use case.

In the forthcoming v0.12.0 release of Terraform we will have a `-json` option on `terraform validate` that will allow you to get the error output in a machine-readable way. Do you think that would be a suitable building block for your listing wrapper script?

(If you want to try it out against an experimental configuration, you could try the beta2 release.)

**Comment by pkolyvas** [comment]
@kaizoku0506 Just following up here. Was Martin's solution workable for you?

**Comment by apparentlymart** [comment]
Hi @kaizoku0506,

I didn't hear back from you, so I'm going to close this on the assumption that my last comment (suggesting to use the `terraform validate -json` option, which was released in Terraform v0.12.0 last year) was sufficient to meet the use-case you described, and I'm going to close this just so we can keep the issue tracker focused on active issues.

If `terraform validate -json` was not sufficient to address your need, please feel free to open a new issue describing what parts of that solution aren't working for you, and we can talk there about possible ways to meet those needs. Thanks!
